Freeze! Adds holding people up with guns.

Ryll-Ryll opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Attacking someone with a gun at point blank with grab intent will now hold them at gunpoint. This means you will automatically shoot the target if they move or attack, though they're still free to rummage around in their (or your, if you continue standing next to them) inventory, speak (including using radios), and throw things. The shooter may roam around a 3 tile box radius without breaking the hold-up, allowing them to reposition themselves, though breaking visibility will break the hold-up. In addition, if you're several tiles away and they make a break for it, they may be able to dodge the projectile. Tactical choices!

[![example.png](https://i.imgur.com/bA4npHs.png)](https://i.imgur.com/bA4npHs.png)
_ABOVE PICTURE IS OUTDATED, UPDATED TO HAVE 2 TILES OF ROAM RANGE INSTEAD OF 3_

For the first 3 seconds after the hold-up, any triggered reaction shot will deal normal damage.
Between 3-10 seconds after, any triggered reaction shot will deal double damage.
After 10 seconds, the reaction shot will deal 2.5x damage.

The ramp-up is to give weapons with weaker per-shot damage like disablers, laser guns, and the .38 the ability to be useful for arrests and stick ups, without affecting combat balance too much. If you got someone you're trying to kill to hold still for 2 seconds at point blank range, you may as well just keep shooting them rather than fiddle with holding them up. If you're gonna resist, it's best to do it quickly!

Lastly, if the shooter is attacked and takes damage from anything, they have a chance to flinch and fire in reaction. This is a 50% chance, unless they're hit in the arm holding the gun, in which case it's an 80% chance.

**(Update 11/10/19)**
Adds an achievement, "Disk, Please!", for being a nuke op and holding up the Captain with a loaded rocket launcher. The captain must be conscious for it to count!

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Adds in mechanical backing for taking people hostage or being able to arrest people with the threat of violence rather than actually carrying out that violence. Trying to take hostages or hold people up right now will just get you immediately shove spammed mid-"DROP $500 OR FAILRP" unless the other person feels like playing along. This makes cooperating or risking getting shot an actual choice, increasing mechanical depth while also adding more roleplay potential.

Also yes, the target cuffing themselves will not trigger the reaction shot, so holding someone up, backing up, and throwing cuffs at them is a valid move. 

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: Ryll/Shaps
add: You can now hold people at gunpoint! Attack someone at point blank with a gun on GRAB intent to initiate the stickup. You will automatically fire at your target if they move or attack, though they can still rummage through their (or your!) inventory, speak (including on the radio), and throw things. Go let loose your inner Omar!
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

actioninja commented:
That's a big box.
I'm not going to preemptively go "waah op" but I got a good feeling people are going to do weird shit with it.

Shapsy commented:
There's a possibility, but you still have to be point blank to initiate it, and the further away you are, the more chance the target has to juke the shot. At 3 tiles away with no slowdown it's definitely possible to dodge given room to maneuver

Couls commented:
god I want to say you should immobilize the target so they don't move immediately and trigger it but I also don't think adding an instant stun move to guns is a good idea

Couls commented on code/datums/components/gunpoint.dm:
```diff
@@ -0,0 +1,103 @@
+	weapon.process_fire(target, shooter)
+	qdel(src)
+
+datum/component/gunpoint/proc/cancel()
```
/datum

Couls commented on code/datums/components/gunpoint.dm:
```diff
@@ -0,0 +1,103 @@
+	to_chat(target, "<span class='userdanger'>[shooter] breaks [shooter.p_their()] aim on you!</span>")
+	qdel(src)
+
+datum/component/gunpoint/proc/flinch(damage, damagetype, def_zone)
```
/datum

Couls reviewed (commented):
some nitpicks

Qustinnus commented on code/datums/components/gunpoint.dm:
```diff
@@ -0,0 +1,103 @@
+	shooter.visible_message("<span class='danger'>[shooter] aims [weapon] point blank at [target]!</span>", \
+		"<span class='danger'>You aim [weapon] point blank at [target]!</span>", target)
+	to_chat(target, "<span class='userdanger'>[shooter] aims [weapon] point blank at you!</span>")
+	target.emote("gasp") // *extremely bad russian accent* no!
```
this is dumb, it takes away the ability for players to reply to the action themselves because you force them to gasp, and this'll cause shit like Ian to gasp because you're holding them at gunpoint.

carlarctg replied:
1. no??? 2. feature

add a MOB_ORGANIC check

Qustinnus commented on code/datums/components/gunpoint.dm:
```diff
@@ -0,0 +1,103 @@
+	if(!can_see(shooter, target, GUNPOINT_SHOOTER_STRAY_RANGE))
+		cancel()
+
+/datum/component/gunpoint/proc/nap_violation()
```
stop making shitty proc names that people have to dig for to understand what they mean, goofball 2.0

wesoda25 replied:
`/datum/component/gunpoint/proc/ripmacmiller()`

Tlaltecuhtli commented:
time to gunpoint the captain with a rocket launcher

Shapsy commented:
Cheers @Couls 

@Qustinnus the gasp was supposed to be an audio cue that the hold-up was taking place a la MGS2/MGS3, but i forgot those got removed cause people didn't like them, so I'll come up with another cue

CRITAWAKETS commented:
Have the shooter say out loud "FREEZE!" as a cue.

Also it would be a good idea to make it require a tiny bit of time before triggering. Like, 0.25 seconds so you cant just FREEZE! someone while they are running perpendicular to you.

Shapsy commented:
I'd rather do like either an alert sound or a gun click/shifting around than verbal speech

Also, it doesn't matter if you hold someone up when they're running past you because you'll just immediately shoot them point blank anyway like you would have without hold ups. I'm not adding a stun for it because it's intended to be used on someone you either surprised (walking up to someone on a console and pulling a pistol) or someone's who's temporarily incapacitated (flashbanged or stamcritted for example),

cacogen commented:
's cool
we bay now

cacogen commented:
's cool
we bay now

jdawg1290 commented:
tie it to aggro grab instead of passive

Shapsy commented:
You're not actually grabbing them at all, the point is that you're just walking up to them and sticking a gun in their face, not that you put them in a chokehold and then stick the gun to them

Couls commented:
Taking hostages should be a thing too imo but thats out of scope for this one

Shapsy commented:
Well a hold up and taking someone hostage basically boils down to the same thing, pointing a gun at someone to restrict their movement, so I don't see why it'd be out of scope. Unless you mean like a meatshield Gears of War style, but I think a neck grab is basically that anyway

Couls commented:
neck grab still hits the holder I think unfortunately

SkeletalElite commented:
What happens if you manually shoot by clicking while someone is held up. Does the shot fire toward the person you have held up even if you click somewhere else? Does a manual shot still have the damage bonus?

Shapsy commented:
You can still fire manually at your locked target/other targets without breaking the hold up, and the damage boost only applies to the reaction shot. The only things that break hold ups are

Triggers shot:
* Target moving
* Target attacking anything with hand or item
* Target shooting a gun
* Shooter has a chance to flinch when hit

Doesn't trigger shot:
* Shooter breaks LoS
* Shooter leaves radius
* Shooter drops or stows their gun

Reloading and otherwise manipulating your gun is a-okay as long as you don't drop it. You can also hold people up with an empty gun! Feelin' lucky, punk?

optimumtact commented:
lmao bay feature

optimumtact commented:
next add a blink button

Shapsy commented:
I STOLE IT FROM GOON

gonna add a fart emote to this next if you don't watch it

Farquaar commented:
Does the shooter still face the target even if he moves? If not, they should.

Shadowflame909 commented:
The only way to make this PR better would be forcing the person in front of you to do *emotes.

Thus, FREEZE! EVERYBODY CLAP YOUR HANDS!

Pugie commented:
This can be used to fire weapons that require 2 hands (like most shotguns) even when the off hand isn't free.

Shapsy commented:
Oh, good looking out, I'll get that

optimumtact commented:
remove the damage ramp up, it doesn't make any sense, auto aim is already super powerful.

Couls commented:
I feel like without the damage ramp up people won't really mind tanking the first shot when held up and it takes a bit of the strength of holding someone up in the first place, considering it already takes a while for the damage to rise I don't think it's all that bad

Shapsy commented:
I agree that the damage ramp up is important to the feature, so that things other than shotguns and .357's have coercive power. And, of course, if someone has you standing still for 10 seconds with a shotgun at close range, they could have killed you anyway.

optimumtact commented:
I don't know, just seems like people will choose death anyway because of cloning

optimumtact commented:
anyway okay I guess, i'll probably regret it later

Couls commented:
At least we can say we gave it a shot

Akrilla commented:
If the first 2 seconds has normal damage, wouldn't it always be better for the hostage to just tank the shot, since the risk of double damage is too much to just wait around for?

Couls commented:
Yeah If you're going to fight back you should do it immediately but you probably won’t do it in situations where you’re not prepared or surprised

Pugie commented:
> If the first 2 seconds has normal damage, wouldn't it always be better for the hostage to just tank the shot, since the risk of double damage is too much to just wait around for?

Probably not. Just shoving someone twice to stun them and then pulling a gun is enough to get the 2*damage ramp ups. For nearly all firearmd this is enough to achieve damage slowdown in 1 shot so the hostage will be at a disadvantage nearly always unless they have an instant stun weapon in melee range, a teleport or a heavy bola.

This does actually make me think though. Can this be used to get around hulk and carp's respective firearm limitations? I'll probably check later today.

Shapsy commented:
I'll fix the corner cases with heavy weapons and non weapon friendly traits in a mo'

ninjanomnom commented on code/datums/components/gunpoint.dm:
```diff
@@ -0,0 +1,108 @@
+	if(!isliving(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	shooter = parent
```
No, cast your parent when you need it, dont store it an extra time

ninjanomnom commented on code/datums/components/gunpoint.dm:
```diff
@@ -0,0 +1,108 @@
+/datum/component/gunpoint/Destroy(force, silent)
+	qdel(status_hold_up)
+	qdel(status_held_up)
+	return ..()
```
You need to null those other references

Shapsy commented:
ugh took care of a bunch of the corner cases with people firing guns that shouldn't be fireable and also made it so bumping or grabbing the target cancels the hold up (i'd rather not deal with balance implications of being able to drag around hostages at the moment), but I need to add a new signal to guns for when they're fired, which I'll do tomorrow

Couls commented:
Shooting people should cancel the holdup so you can't shoot people to bait them into moving

Shapsy commented:
If the person performing the hold-up is shooting you to bait you into moving, chances are they were just gonna kill you anyway

Couls commented:
In that case they shouldn't really get the damage ramp-up all things considered

SomeguyManperson commented on code/datums/components/gunpoint.dm:
```diff
@@ -0,0 +1,145 @@
+	var/mob/living/shooter = parent
+	target = targ
+	weapon = wep
+	RegisterSignal(targ, list(COMSIG_MOB_ATTACK_HAND, COMSIG_MOB_ITEM_ATTACK, COMSIG_MOB_ATTACK_RANGED, COMSIG_MOVABLE_MOVED), .proc/trigger_reaction)
```
Double check that the MOB_ATTACK_RANGED signal doesn't trigger if the person unarmed clicks on a tile they can't reach. From what I can see the signal doesn't send if they fired a gun, but does if they attempted to unarmed click outside their melee range, which only works as intended here if they tried initiating (but apparently not throwing things with) telekinesis. Might also want to make throwing trigger the shot if throwing stuff is considered lethal enough to matter.

Shapsy replied:
blorble, I realized the issue with MOB_ATTACK_RANGED and thought I pushed removing it, but yeah I will probably follow your approach with adding a signal onto guns firing

also I'm gonna leave throwing unchecked since there can be legit reasons for it ("throw me your ID card!"), and any damage is a 50/50 chance to cause a flinch anyway so it works out as a tactical choice

Also re: cancelling if the gunman fires a gun, I'd rather not do that since it'd severely limit the hostage taking potential if other people could rush in knowing the gunman loses their edge if they fire. If it's problematic, I can change it afterwards

Shapsy commented:
Re: cancelling if the gunman fires a gun, I'd rather not do that since it'd severely limit the hostage taking potential if other people could rush in knowing the gunman loses their edge if they fire. If it's problematic, I can change it afterwards

Couls commented:
That’s understandable

Shapsy commented:
Alright this should be good to go!

Shapsy commented:
uh

Ryll-Ryll committed:
there's a time to talk, and there's a time to shut up


now is the time to be quiet, son


newlines, trigger on ranged attacks


slash


new cheevie


oop


my nice neat code ;;;_;;;


slightly less bad


flinch

Shapsy commented:
Hedging my bets on this cause I don't want to dive right into balance issues

-Range is now 2 tiles of roaming
-Ramp up is 3 seconds to 2x instead of 2 seconds, still 10 seconds to 2.5x

Bastian0930 commented:
Shit litterally crashes the master controller, at least on beestation. We hit all sorts of defcons, makin it look like a math equation. How you can do it is by aiming like 100 times at someone, with them also aiming at you. Then, shoot the gun or whatever, and you kill the server.

Shadowflame909 commented:
>Two people holding each other up 100 times creates a paradox that crashes reality.

Sounds like an SNL Skit. (Mmm' Watcha Say...)

Shapsy commented:
oops!!! I thought for sure I had accounted for that, I'll push a fix out for it in the next hour and a half. Thanks for doing our testing!

cacogen commented:
its like the end of reservoir dogs mixed with the end of the sopranos

Couls commented:
wow two spoilers in one

optimumtact commented:
what is the status of this?

Shapsy commented:
Still ready to go, but actually now that achievements are a thing again let me uncomment the achieveo

optimumtact commented:
watch out for #48007

Shapsy commented:
fixed a few more issues, changed stage 2 to be reached in 2.5 seconds, between 2 and 3, i'm satisfied with this now

Alexiusgaiusceasar commented:
If the gun is in Person A's inactive hand, and Person B triggers one of the responses that causes a shot to be fired, does it still fire?
